File: android/src/main/java/com/brentvatne/exoplayer/DefaultReactExoplayerConfig.java
package com.brentvatne.exoplayer;

import android.content.Context;

import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultLoadErrorHandlingPolicy;
import com.google.android.exoplayer2.upstream.LoadErrorHandlingPolicy;

public class DefaultReactExoplayerConfig implements ReactExoplayerConfig {

    private final DefaultBandwidthMeter bandwidthMeter;
    private boolean disableDisconnectError = false;

    public DefaultReactExoplayerConfig(Context context) {
        this.bandwidthMeter = new DefaultBandwidthMeter.Builder(context).build();
    }

    public LoadErrorHandlingPolicy buildLoadErrorHandlingPolicy(int minLoadRetryCount) {
        if (this.disableDisconnectError) {
            // Use custom error handling policy to prevent throwing an error when losing network connection
            return new ReactExoplayerLoadErrorHandlingPolicy(minLoadRetryCount);
        }
        return new DefaultLoadErrorHandlingPolicy(minLoadRetryCount);
    }

    public void setDisableDisconnectError(boolean disableDisconnectError) {
        this.disableDisconnectError = disableDisconnectError;
    }

    public boolean getDisableDisconnectError() {
        return this.disableDisconnectError;
    }

    @Override
    public DefaultBandwidthMeter getBandwidthMeter() {
        return bandwidthMeter;
    }
}

File: android/src/main/java/com/brentvatne/exoplayer/DefaultReactExoplayerConfig.kt
package com.brentvatne.exoplayer

import android.content.Context
import androidx.media3.exoplayer.upstream.DefaultBandwidthMeter
import androidx.media3.exoplayer.upstream.DefaultLoadErrorHandlingPolicy
import androidx.media3.exoplayer.upstream.LoadErrorHandlingPolicy

class DefaultReactExoplayerConfig(private val context: Context, override var initialBitrate: Long? = null) : ReactExoplayerConfig {

    private var bandWidthMeter: DefaultBandwidthMeter = createBandwidthMeter(initialBitrate)

    override var disableDisconnectError: Boolean = false

    override val bandwidthMeter: DefaultBandwidthMeter
        get() = bandWidthMeter

    private fun createBandwidthMeter(bitrate: Long?): DefaultBandwidthMeter =
        DefaultBandwidthMeter.Builder(context)
            .setInitialBitrateEstimate(bitrate ?: DefaultBandwidthMeter.DEFAULT_INITIAL_BITRATE_ESTIMATE)
            .build()

    override fun setInitialBitrate(bitrate: Long) {
        if (initialBitrate == bitrate) return
        initialBitrate = bitrate
        bandWidthMeter = createBandwidthMeter(bitrate)
    }

    override fun buildLoadErrorHandlingPolicy(minLoadRetryCount: Int): LoadErrorHandlingPolicy =
        if (disableDisconnectError) {
            ReactExoplayerLoadErrorHandlingPolicy(minLoadRetryCount)
        } else {
            DefaultLoadErrorHandlingPolicy(minLoadRetryCount)
        }
}
